crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -353,13 +525,42 @@ where
     }
 
     fn transaction_by_hash(&self, hash: TxHash) -> ProviderResult<Option<TransactionSigned>> {
+        for block_state in self.canonical_in_memory_state.canonical_chain() {
+            if let Some(tx) = block_state.block().block().body.iter().find(|tx| tx.hash() == hash) {
+                return Ok(Some(tx.clone()));
+            }
+        }
+
         self.database.transaction_by_hash(hash)
     }
 
     fn transaction_by_hash_with_meta(
         &self,
         tx_hash: TxHash,
     ) -> ProviderResult<Option<(TransactionSigned, TransactionMeta)>> {
+        for (block_number, block_state) in
+            self.canonical_in_memory_state.canonical_chain().enumerate()
+        {
```

Comment:
I'd actually like to move these helper functions into the canonical memory type directly

Reply:
ok, I've moved the tx-related functions to `CanonicalInMemoryState`, should we move anything else?